### ISSUE TITLE
Duplicated unused bytecodes at end of function

### ISSUE THREAD

**Issue by nedbat** [python/cpython#86862]
BPO | [42696](https://bugs.python.org/issue42696)
--- | :---
Nosy | @rhettinger, @nedbat, @markshannon

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/markshannon'
closed_at = None
created_at = <Date 2020-12-20.21:29:18.434>
labels = ['3.10']
title = 'Duplicated unused bytecodes at end of function'
updated_at = <Date 2020-12-22.20:46:01.201>
user = 'https://github.com/nedbat'
```

bugs.python.org fields:
```python
activity = <Date 2020-12-22.20:46:01.201>
actor = 'Mark.Shannon'
assignee = 'Mark.Shannon'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2020-12-20.21:29:18.434>
creator = 'nedbat'
dependencies = []
files = []
hgrepos = []
issue_num = 42696
keywords = []
message_count = 6.0
messages = ['383460', '383510', '383512', '383598', '383602', '383613']
nosy_count = 3.0
nosy_names = ['rhettinger', 'nedbat', 'Mark.Shannon']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue42696'
versions = ['Python 3.10']
```

</p></details>

**Comment by iritkatriel** [comment]
TL;DR:   the ``return`` code is no longer duplicated for this example (but if you remove the ``return 17`` then the implicit ``return None`` is still duplicated).

The long story:

The reason for the duplication is a small-exit-block-inlineing heuristic, where if we have an unconditional jump to an exit block of at most 4 instructions, that exit block is inlined and the jump is eliminated. I recently disabled this optimisation for the case where the exit block is associated with a line number, because it caused issues for the debugger (see https://github.com/python/cpython/pull/94592). 

For cases where the block doesn't have a line number (as in the case of the implicit ``return None``), we need to duplicate the block because the ``return`` needs to be traced with the line number of the last instruction that was executed that has a line number. Since it's too costly to keep track of this line number at runtime, the line info is static and this means that we need a copy of the ``return None`` for each location that it can be reached from.

Not duplicated:

```
>>> def f():
...         for i in range(10):
...             break
...         return 17
... 
>>> import dis
>>> dis.dis(f)
  1           0 RESUME                   0

  2           2 LOAD_GLOBAL              1 (NULL + range)
             14 LOAD_CONST               1 (10)
             16 CALL                     1
             26 GET_ITER
             28 FOR_ITER                 2 (to 36)
             32 STORE_FAST               0 (i)

  3          34 POP_TOP

  4     >>   36 LOAD_CONST               2 (17)
             38 RETURN_VALUE

```

Duplicated:

```
>>> def g():
...         for i in range(10):
...             break
... 
>>> dis.dis(g)
  1           0 RESUME                   0

  2           2 LOAD_GLOBAL              1 (NULL + range)
             14 LOAD_CONST               1 (10)
             16 CALL                     1
             26 GET_ITER
             28 FOR_ITER                 4 (to 40)
             32 STORE_FAST               0 (i)

  3          34 POP_TOP
             36 LOAD_CONST               0 (None)
             38 RETURN_VALUE

  2     >>   40 LOAD_CONST               0 (None)
             42 RETURN_VALUE

```
